Feature/cmcd settings from mpd and steering report

matiasrb97 opened the pull request:
This PR aims to add two features related to CMCD.

First one is CMCD session initialization and configuration through the MPD as specified in the DASH 6th draft version standard.
Second one is the ability to send CMCD data to the Content Steering server.

Summary of changes:

- Parse the new  `ClientDataReporting ` and `CMCDParameters` settings from mpd file following the DASH specification.
- Implement the functionality to use the new parsed settings: `serviceLocations` , `adaptationSets` , `includeInRequests` filters, `mode` ,` version` , `keys` , and other configurations for  `CMCD` specified in the DASH 6th draft version standard.
- Implemented `_getCmcdDataForSteering` for sending CMCD data to content steering server.
- Implemented a set of logs when required settings are not configured in the MPD for CMCD, or when the settings have wrong values. e.g: `<CMCDParameters version='1' mode='query' keys='foo bar'>` in this example foo bar are not valid values.
- Created tests for both new features.

Note:
If `applyCMCDParameters`  is `true`, the player will use the `CMCD` settings from the MPD if specified, ignoring whatever the player's own settings for `CMCD` .
If `applyCMCDParameters`  is `false`, the player won't use the `CMCD` settings from the MPD ignoring whatever the MPD has set for `CMCD`.

This feature can be tested using this simple testing tool: https://github.com/qualabs/cmcd-content-steering-test-runner.
It has a set of use cases which we used for developing and testing the features implemented in this work.

List of use cases tested:
Case 1: Basic MPD Configuration. - _It has a simple configuration of CMCD._
Case 2: Basic MPD Configuration with Service Locations filter.
Case 3: Enable CMCD From JS Code. - _If "CMCD from JS" checkbox is false, CMCD will not be sent._
Case 4: Multiple ServiceDescription elements.  - _It will use the latest ServiceDescription element._
Case 5: Multiperiod MPD with ServiceDescription. - _We kept this case out of scope in this work, it implies other changes not related to just cmcd._
Case 6: Setting serviceLocations & includeInRequest filters.
Case 7: Setting includeInRequests only filter.
Case 8: Setting serviceLocations & includeInRequests filters.
Case 9: includeInRequest default value.
Case 10: Keys are not defined.
Case 11: Keys parameter has some invalid key.
Case 12: Keys are defined but all the keys are not implemented.
Case 13: includeInRequests is defined but has some invalid request type.
Case 14: includeInRequests are defined but all the request types are not implemented.
Case 15: Sending CMCD data to the steering server from the last video request.
Case 16: Sending CMCD data to the steering server from the last audio request if MPD has only audio tracks.

In each case you should be able to see the CMCD sent in each request and in the bottom the MPD manifest used in the selected use case.

dsilhavy commented on src/streaming/models/CmcdModel.js:
```diff
@@ -191,28 +212,150 @@ function CmcdModel() {
     }
 
+    function isCmcdEnabled() {
+        const cmcdParameters = getCmcdParametersFromManifest();
```
Please rename `cmcdParameters` to `cmcdParametersFromManifest`

dsilhavy commented on src/streaming/models/CmcdModel.js:
```diff
@@ -191,28 +212,150 @@ function CmcdModel() {
+        return cmcdParameters.version ? true : settings.get().streaming.cmcd && settings.get().streaming.cmcd.enabled;
+    }
+
+    function _checkIncludeInRequests(cmcdParameters) {
```
Please rename `cmcdParameters` to `cmcdParametersFromManifest`

dsilhavy commented on src/streaming/models/CmcdModel.js:
```diff
@@ -191,28 +212,150 @@ function CmcdModel() {
+        const enabledCMCDKeys = cmcdParameters.version ? cmcdParameters.keys.split(' ') : settings.get().streaming.cmcd.enabledKeys;
+        const invalidKeys = enabledCMCDKeys.filter(k => !defaultAvailableKeys.includes(k));
+
+        if (invalidKeys.length == enabledCMCDKeys.length && enabledCMCDKeys.length > 0) {
```
Use strong equality `===`

dsilhavy replied:
Please address this

dsilhavy commented on src/streaming/models/CmcdModel.js:
```diff
@@ -45,29 +46,44 @@ import {encodeCmcd} from '@svta/common-media-library/cmcd/encodeCmcd';
 import {toCmcdHeaders} from '@svta/common-media-library/cmcd/toCmcdHeaders';
 
 const CMCD_VERSION = 1;
+const CMCD_ALL_REQUESTS = '*';
```
Please rename to `CMCD_ALL_REQUESTS_PLACEHOLDER` or sth more descriptive

dsilhavy commented on src/streaming/models/CmcdModel.js:
```diff
@@ -191,28 +212,150 @@ function CmcdModel() {
 
+    function _updateLastMediaTypeRequest(type, mediatype) {
+        // Video > Audio > None
+        if (mediatype == Constants.VIDEO || mediatype == Constants.AUDIO) {
```
Use strong equality `===`

dsilhavy commented on src/streaming/models/CmcdModel.js:
```diff
@@ -348,13 +492,17 @@ function CmcdModel() {
         let cid = settings.get().streaming.cmcd.cid ? settings.get().streaming.cmcd.cid : internalData.cid;
 
-        data.v = CMCD_VERSION;
+        data.v = cmcdParameters.version ? cmcdParameters.version : CMCD_VERSION;
```
In my opinion we can not simply take the version from the MPD here. We need to use the version that we use to generate the reporting fields.  In our case that is always version 1 atm. The `version` as specified in the MPD only tells use the _ highest CMCD version as defined in CTA-5004 that is accepted by the reporting server._

matiasrb97 replied:
@dsilhavy Thanks for your review and comments!! I did an update with the first set of changes and I'm gonna left questions in some of your comments for more details or to check if I understood well. There are changes that I didn't make yet because I'll need to clarify. 

In this case, is it okay to left `data.v = CMCD_VERSION` then?

dsilhavy replied:
Thanks @matiasrb97 I am checking. Yes let's keep `data.v = CMCD_VERSION` here.

dsilhavy commented on src/streaming/models/CmcdModel.js:
```diff
@@ -348,13 +492,17 @@ function CmcdModel() {
         data.sid = settings.get().streaming.cmcd.sid ? settings.get().streaming.cmcd.sid : internalData.sid;
 
+        cid = cmcdParameters.contentID ? cmcdParameters.contentID : cid;
+        data.sid = cmcdParameters.sessionID ? cmcdParameters.sessionID : data.sid;
```
Move this part up to the assignment of sid

dsilhavy commented on src/streaming/net/HTTPLoader.js:
```diff
@@ -566,9 +569,13 @@ function HTTPLoader(cfg) {
-
-        if (settings.get().streaming.cmcd && settings.get().streaming.cmcd.enabled) {
-            const cmcdMode = settings.get().streaming.cmcd.mode;
+        const currentServiceLocation = request?.serviceLocation;
```
Please move the whole CMCD logic to a dedicated function that is called within `_updateRequestUrlAndHeaders`

dsilhavy replied:
If this is too much work let me know then I do this in a dedicated PR. Thanks

dsilhavy commented on test/unit/dash.models.DashManifestModel.js:
```diff
@@ -1508,5 +1508,92 @@ describe('DashManifestModel', function () {
+                }
+                const data = dashManifestModel.getServiceDescriptions(manifestData);
+                const clientDataReporting = data[0].clientDataReporting;
+                expect(clientDataReporting.serviceLocations).to.be.equal('cdn-a cdn-b');
```
Please also check the asArray logic here if possible

dsilhavy commented on test/unit/dash.models.DashManifestModel.js:
```diff
@@ -1508,5 +1508,92 @@ describe('DashManifestModel', function () {
+                        'ClientDataReporting':{
+                            'CMCDParameters': {
+
+                            },
```
Please also check for the right default values here for `mode` and `includeInRequests`

dsilhavy commented:
@matiasrb97 Thank you very much for your work on this. I added some review comments and some additional comments and questions below. I am discussing them with the authors on Slack but posting here as well: 

* Checking that with the authors of the spec: `ReportingSystem` has type `DescriptorType` . Does that imply that `CMCDParameters` as defined in TableK.4.2.7.2-1 needs to have a `schemeIdUri` as well (`urn:mpeg:dash:cta-5004:2023`)? We are currently not checking for this schemeId: `urn:mpeg:dash:cta-5004:2023` 
* Are we checking for multiple `ReportingSystem` (`CMCDParameters`) elements in `ClientDataReporting`
* TableK.4.2.7.2-1 says "Semantics of CMCD Reporting element" while the element itself is named "CMCDParameters"
* ClientDataReportingType can have a sequence/multiple ReportingSystem elements. According to Table K.4.2.7.1-1 only one CMCDParameters element is possible. Is that correct?
* For CMCDParameterType : According to Table K.4.2.7.2-1 version is mandatory which does not seem to be reflected in the schema definition: <xs:attribute name="version" type="xs:unsignedInt"/> . Do we need to add use="required"
* The default value for includeInRequests seems to differ :
  * TableK.4.2.7.2-1: segments
  * Schema: <xs:attribute name="includeInRequests" type="xs:string" default="*"/>
* <xs:attribute name="keys" type="StringNoWhitespaceVectorType"/> is mandatory according to TableK.4.2.7.2-1

dsilhavy commented:
Some things I forgot:

* It would be nice to have a sample page similar to this one for CMCD triggered via the MPD: https://reference.dashif.org/dash.js/nightly/samples/advanced/cmcd.html
* Can you share a publicly hosted MPD (maybe one of these https://github.com/qualabs/cmcd-content-steering-test-runner/tree/main/dash-cases/cases) that we include in the reference UI? I will also use this to implement a functional test then.

ZmGorynych commented:
> @matiasrb97 Thank you very much for your work on this. I added some review comments and some additional comments and questions below. I am discussing them with the authors on Slack but posting here as well:
> 
> * Checking that with the authors of the spec: `ReportingSystem` has type `DescriptorType` . Does that imply that `CMCDParameters` as defined in TableK.4.2.7.2-1 needs to have a `schemeIdUri` as well (`urn:mpeg:dash:cta-5004:2023`)? We are currently not checking for this schemeId: `urn:mpeg:dash:cta-5004:2023`
Yes.
> * Are we checking for multiple `ReportingSystem` (`CMCDParameters`) elements in `ClientDataReporting`
> * TableK.4.2.7.2-1 says "Semantics of CMCD Reporting element" while the element itself is named "CMCDParameters"
> * ClientDataReportingType can have a sequence/multiple ReportingSystem elements. According to Table K.4.2.7.1-1 only one CMCDParameters element is possible. Is that correct?
Multiple should be allowed. You can report different keys to different endpoints. Imagine only sending cid and sid to content steering server to let the backend correlate with what it sees from the logs.

> * For CMCDParameterType : According to Table K.4.2.7.2-1 version is mandatory which does not seem to be reflected in the schema definition: <xs:attribute name="version" type="xs:unsignedInt"/> . Do we need to add use="required"
Yes.

> * The default value for includeInRequests seems to differ :
> 
>   * TableK.4.2.7.2-1: segments
>   * Schema: <xs:attribute name="includeInRequests" type="xs:string" default="*"/>
> * <xs:attribute name="keys" type="StringNoWhitespaceVectorType"/> is mandatory according to TableK.4.2.7.2-1

`<xs:attribute name="includeInRequests" type="xs:string" default="*"/>` for CMCD is madness, it should be segments

matiasrb97 commented on src/dash/models/DashManifestModel.js:
```diff
@@ -1255,6 +1256,26 @@ function DashManifestModel() {
+        const entry = new ClientDataReporting();
+
+        if (element.hasOwnProperty(DashConstants.CMCD_PARAMETERS)) {
+            entry.CMCDParameters = element[DashConstants.CMCD_PARAMETERS];
```
@dsilhavy Here is where I think I didn't understand your previous comment about creating a dedicated model class for `CMCDParameters`. 

Let me see if I'm understanding what you are saying: 
- Create a vo class for `CMCDParameters` and check for the mandatory fields when initializing it.

I already created the class in my latest changes, so I have to put the checks for the mandatory fields now.

matiasrb97 commented on src/streaming/models/ClientDataReportingModel.js:
```diff
@@ -0,0 +1,74 @@
+import FactoryMaker from '../../core/FactoryMaker.js';
+import {HTTPRequest} from '../vo/metrics/HTTPRequest.js';
+
+function ClientDataReportingModel() {
```
@dsilhavy So, you are saying that we should change `ClientDataReportingModel.js` to `ClientDataReportingController.js` (and also move it to the `controllers` folder I guess), is that correct?

dsilhavy replied:
Yes, would be great if you can change it like this. Change `ClientDataReportingModel.js` to `ClientDataReportingController.js`.

matiasrb97 commented on src/streaming/models/CmcdModel.js:
```diff
@@ -191,28 +212,150 @@ function CmcdModel() {
+                return false;
+            }
+        }
+        return cmcdParameters.version ? true : settings.get().streaming.cmcd && settings.get().streaming.cmcd.enabled;
```
@dsilhavy  As `version` is mandatory field, we used it to check if we should use manifest settings above JS settings. But you are correct, on previous `if` statement we are checking for `version` and `keys` so I'll improve this line to remove checking for `version` again.

dsilhavy commented on src/streaming/models/CmcdModel.js:
```diff
@@ -191,28 +202,149 @@ function CmcdModel() {
+        let enabledRequests = settings.get().streaming.cmcd.includeInRequests;
+
+        if (cmcdParametersFromManifest.version) {
+            enabledRequests = cmcdParametersFromManifest.includeInRequests ?? [CMCD_ALL_REQUESTS_PLACEHOLDER];
```
The default value for `includeInRequests` shall be `segments`. See also the comment from @ZmGorynych

dsilhavy commented on src/dash/vo/CMCDParameters.js:
```diff
@@ -0,0 +1,63 @@
+            this.contentID = data.contentID;
+            this.mode = data.mode;
+            this.keys = data.keys ? data.keys.split(' ') : null;
+            this.includeInRequests = data.includeInRequests ? data.includeInRequests.split(' ') : null;
```
Can we set the default values here directly? The default value for `mode` is "query" and the default value for `includeInRequests` is "segments"

dsilhavy commented:
@matiasrb97 Thank you for addressing most of my comments. I did another quick review, please see the comments above. One additional item based on the feedback by @ZmGorynych : Can we add a check for the `urn:mpeg:dash:cta-5004:2023` scheme id before creating an instance of `CMCDParameters`.

matiasrb97 commented:
> @matiasrb97 Thank you for addressing most of my comments. I did another quick review, please see the comments above. One additional item based on the feedback by @ZmGorynych : Can we add a check for the `urn:mpeg:dash:cta-5004:2023` scheme id before creating an instance of `CMCDParameters`.

@dsilhavy Thanks again for your comments! I think that now I finished addressing all your comments at the moment. Please check and maybe we can resolve all the comments that are closed now _(and keep the ones that need something else)_. 

Regarding the schemeId, can you check please if I understood well ? I did the change in this single [commit](https://github.com/Dash-Industry-Forum/dash.js/pull/4426/commits/4a010711fa03e946853460a61c0bf430655d8ccc) so it's easy to revert if I missunderstood.

dsilhavy commented:
@matiasrb97 This looks all good to me. Is there anything left from your side? If not, can you please merge the changes from development one more time then I can merge your PR.

matiasrb97 commented:
> @matiasrb97 This looks all good to me. Is there anything left from your side? If not, can you please merge the changes from development one more time then I can merge your PR.

Great, thanks! Nothing else from my side, I updated this branch with the latest from development. 

And regarding the hosted mpd that you asked for, here I have two links with two mpd samples that are used in the testing repository: 
- https://d23x4gyjtx25iu.cloudfront.net/1.mpd
- https://d23x4gyjtx25iu.cloudfront.net/8.mpd
